File: backend/utils/ticker.py
import re

from config import Settings
from utils.exceptions import InvalidTickerError

_TICKER_PATTERN = re.compile(r"^[A-Za-z0-9.\-^=]+$")


def normalize_ticker(raw_ticker: str) -> str:
    return raw_ticker.strip().upper()


def validate_ticker(raw_ticker: str, settings: Settings) -> str:
    ticker = normalize_ticker(raw_ticker)

    if not ticker:
        raise InvalidTickerError("Ticker symbol is required.", ticker=raw_ticker)

    if (
        len(ticker) < settings.ticker_min_length
        or len(ticker) > settings.ticker_max_length
    ):
        raise InvalidTickerError(
            (
                f"Ticker must be between {settings.ticker_min_length} "
                f"and {settings.ticker_max_length} characters."
            ),
            ticker=ticker,
        )

    if not _TICKER_PATTERN.match(ticker):
        raise InvalidTickerError(
            "Ticker contains invalid characters.",
            ticker=ticker,
        )

    return ticker

File: frontend/src/utils/ticker.js
import { TICKER_MAX_LENGTH } from '../constants'

/** Allowed ticker characters (aligned with backend `utils/ticker.py`). */
const TICKER_PATTERN = /^[A-Za-z0-9.\-^=]+$/

export function normalizeTicker(value) {
  return String(value || '').trim().toUpperCase()
}

/**
 * Validate a ticker for client-side forms.
 * @returns {{ ok: true, ticker: string } | { ok: false, message: string }}
 */
export function validateTickerInput(value) {
  const ticker = normalizeTicker(value)

  if (!ticker) {
    return { ok: false, message: 'Ticker symbol is required.' }
  }

  if (ticker.length > TICKER_MAX_LENGTH) {
    return {
      ok: false,
      message: `Ticker must be at most ${TICKER_MAX_LENGTH} characters.`,
    }
  }

  if (!TICKER_PATTERN.test(ticker)) {
    return {
      ok: false,
      message: 'Ticker contains invalid characters. Use letters, numbers, and . - ^ = only.',
    }
  }

  return { ok: true, ticker }
}
